### ISSUE TITLE
Passing merge strategy options to the 'rebase' command fails due to unescaped quotes

### ISSUE THREAD

**Issue by sonntam** [git-for-windows/git#1321]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.14.2.windows.2
built from commit: 34c1a976dcdaa107499f5c0c1c416bcfee477f94
sizeof-long: 4
machine: x86_64

```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.15063]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
> type "C:\Program Files (x86)\Git\etc\install-options.txt"

Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Nothing extraordinary

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

The bundeled msys shell

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git rebase --interactive --preserve-merges -m -s recursive --strategy-option=theirs '<commit>^'
```
 - What did you expect to occur after running these commands?

I wanted to change a single commit without changing commit order. I know that --interactive and --preserve-merges are only usable in a limited fashion but in this case it should work. I guess it is some sort of quotes escaping that goes wrong.

 - What actually happened instead?

The merging aborted with an error

**fatal: Unknown option for merge-recursive: -X'theirs'
Error redoing merge <commit>**

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Sorry, it is a private repository. But here is a minimal example to reproduce the error:
```
cd <to-empty-directory>
git init

echo . > test.txt
git add test.txt
git commit -m "initial commit"

git checkout -b testbranch
echo : > test.txt
git commit -am "changed on branch"

git checkout master
echo ; > test.txt
git commit -am "changed on master"

git merge testbranch
echo : > test.txt
git commit -am "merged branch into master"

git rebase --interactive --preserve-merges --strategy-option=theirs 'HEAD~2'
```
now change the first 'pick' to 'edit', then save & quit the editor followed by the commands

```
git commit --no-edit --amend --all
git rebase --continue
< here the error occurs >
```

 The error occurs at a point where a branch is merged into master, i.e. where mergin is actually needed first. The **--strategy-option=theirs** seems to be the cause.

Edit: I checked on my ubuntu computer - there it works flawlessly. So I think it must be shell/windows related.

**Comment by dscho** [comment]
Thank you for your bug report.

> Sorry, it is a private repository

Please note that you now force me to write a script myself to trigger the error by creating a test repository that is *not* private. That immediately makes it impossible for me to work on this, as I have no time for this ☹️

**Comment by PhilipOakley** [comment]
@sonntam Is this a new error that has shown because you updated the git version, or is this just the first time you have seen this error?

There has been some work on updating the `rebase` code and I wonder if this was part of it. ('this' being the re-writing of `--strategy-option=theirs` to `-X'theirs'` )

**Comment by sonntam** [comment]
@dscho Sorry for being lazy. I have added a minimal example to reproduce the error, see above in my original post.

@PhilipOakley  I just ran into it - i have never used that before.

**Comment by PhilipOakley** [comment]
@sonntam Thanks for the confirmation. It's usually better to include the updates in the new post because those who receive email notifications don't see the changes to the old posts..

That said, the little bit at the end that your MVCE works on Ubuntu (but not on Windows) is informative. It suggests to me that the issue is not exactly as per your initial title (`fails due to unescaped quotes`), rather it's that some part of the GfW/MSYS2 stack is *adding* the quotes. 

That is, I believe that on Linux there would not be any quotes, but we don't actually see the command, maybe we need to add a GIT_TRACE=1 to see the intermediate steps on the two OS's. (https://projects.cs.uaf.edu/redmine/projects/public/wiki/How_do_I_make_Git_give_more_verbose_output for a random internet explanation;-)

**Comment by sonntam** [comment]
@PhilipOakley Thanks for the hint! I tried within the standard windows command window. The same error is thrown there. So I guess it's not a MSYS related issue. 

Here is the relevant part of the trace with `GIT_TRACE = 1`
```
13:55:06.779703 git.c:340               trace: built-in: git 'cat-file' 'commit' '1612649979dbf3c814f1ce89517c744a67af4517'
13:55:06.903486 git.c:340               trace: built-in: git 'merge' '--no-log' '--no-ff' '--strategy=recursive' '-X' ''\''theirs'\''' '-m' 'merged branch into master' '3fe41b165a96f331dc09fdbde544ea247ec32d66'
fatal: Unknown option for merge-recursive: -X'theirs'
13:55:07.085378 git.c:572               trace: exec: 'git-sh-i18n--envsubst' '--variables' 'Error redoing merge $sha1'
13:55:07.085867 run-command.c:626       trace: run_command: 'git-sh-i18n--envsubst' '--variables' 'Error redoing merge $sha1'
13:55:07.215022 git.c:572               trace: exec: 'git-sh-i18n--envsubst' 'Error redoing merge $sha1'
13:55:07.215998 run-command.c:626       trace: run_command: 'git-sh-i18n--envsubst' 'Error redoing merge $sha1'
```

The `'-X' ''\''theirs'\'''` shows that some quotes are added.

**Comment by sonntam** [comment]
This is the output from git 2.1.4

```
14:04:37.858705 git.c:349               trace: built-in: git 'cat-file' 'commit' '4ce36fde964d38461b5a5197adecbbb2a61b90bf'
14:04:37.865068 git.c:349               trace: built-in: git 'merge' '--no-log' '--no-ff' '-m' 'merged branch into master' '84619c006a5f98b4e111aaba19674727b10b259a'
Auto-merging test.txt
CONFLICT (content): Merge conflict in test.txt
Automatic merge failed; fix conflicts and then commit the result.
```

Now I'm intrigued. The trace (thanks for the hint to that, I didn't know about that before) shows that my merge options haven't been passed to the `git merge` command at all! Somehow the Linux version behaves completely different. I tried passing `-m -s recursive` as well - no luck.

**Comment by dscho** [comment]
@sonntam could you have a look at the `.git\rebase-merge` directory after that `edit` step returns control to you? It should have some file with the `-X'theirs'` in it, and I think it may be a combination of `git-rebase` being a shell script and `git rebase--helper --continue` being a MINGW process that gets the unquoting/requoting wrong.

**Comment by sonntam** [comment]
@dscho I did a search and it only has come up with one hit:
```
$ grep -nr theirs .
./rebase-merge/strategy_opts:1:  '--theirs'
```

And that is the whole content of `strategy_opts`, only `'--theirs'` (with quotes)

**Comment by kgybels** [comment]
@dscho I could reproduce this on Ubuntu running git v2.7.4.

The strategy option is rewritten to `-X'theirs'` and stored in `$strategy_args` on [git-rebase--interactive.sh#L89](/git/git/blob/v2.14.2/git-rebase--interactive.sh#L89). This rewritten option is used on [git-rebase--interactive.sh#L394-L397](https://github.com/git/git/blob/v2.14.2/git-rebase--interactive.sh#L394-L397):
```
			if ! do_with_author output eval \
			'git merge ${gpg_sign_opt:+"$gpg_sign_opt"} \
				$allow_rerere_autoupdate $merge_args \
				$strategy_args -m "$msg_content" $new_parents'
```

 Notice how the whole command string to eval is quoted (`eval 'git merge ...'`). This causes the parameter to be passed to the command with the quotes intact.

Rewriting the eval, to be similar to the other ones, fixes the problem:
```
			if ! do_with_author output eval \
			git merge ${gpg_sign_opt:+"$gpg_sign_opt"} \
				$allow_rerere_autoupdate "$merge_args" \
				"$strategy_args" -m "'$msg_content'" "$new_parents"
```

**Comment by dscho** [comment]
@kgybels great find!

I actually got stuck when I tried to figure out a way how to reproduce this in Git's regression test suite. This patch seems to *not* reproduce the issue, even if i thought it should:

```diff
diff --git a/t/t3418-rebase-continue.sh b/t/t3418-rebase-continue.sh
index ca8fb4fb4a2..31f4e4ab872 100755
--- a/t/t3418-rebase-continue.sh
+++ b/t/t3418-rebase-continue.sh
@@ -74,6 +74,22 @@ test_expect_success 'rebase --continue remembers merge strategy and options' '
 	test -f funny.was.run
 '
 
+test_expect_success 'rebase passes merge strategy options correctly' '
+	rm -fr .git/rebase-* &&
+	git reset --hard commit-new-file-F3-on-topic-branch &&
+	test_commit theirs-to-merge &&
+	git reset --hard HEAD^ &&
+	test_commit some-commit &&
+	test_tick &&
+	git merge --no-ff theirs-to-merge &&
+git log --graph --oneline --left-right &&
+	FAKE_LINES="1 edit 2 3" git rebase -i -p -m \
+		-s recursive --strategy-option=theirs HEAD~2 &&
+cat .git/rebase-merge/strategy_opts &&
+	git rebase --continue
+'
+
+exit
 test_expect_success 'setup rerere database' '
 	rm -fr .git/rebase-* &&
 	git reset --hard commit-new-file-F3-on-topic-branch &&

```

**Comment by sonntam** [comment]
@kgybels: That fixes this problem, indeed. Thanks!

**Comment by kgybels** [comment]
>@dscho I actually got stuck when I tried to figure out a way how to reproduce this in Git's regression test suite. 

You need to make sure one of the parents changes, otherwise it will fast-forward instead of redoing the merge (see comment on [git-rebase--interactive.sh#L323](https://github.com/git/git/blob/v2.14.2/git-rebase--interactive.sh#L323) and code on [git-rebase--interactive.sh#L367-L372](https://github.com/git/git/blob/v2.14.2/git-rebase--interactive.sh#L367-L372)). This is also why @sonntam's steps don't reproduce if you leave out `git commit --no-edit --amend --all`. You can turn on verbose for the interactive rebase, then it will tell you when it does a fast-forward.

Here is the result of @sonntam's steps with verbose on, but skipped `git commit --no-edit --amend --all`:
```
$ git rebase --continue
Rebasing (2/3)
d16cc307b80b0cccfbb36e84420a9cc2ceb87f74
Fast-forward to d16cc307b80b0cccfbb36e84420a9cc2ceb87f74
HEAD is now at d16cc30 changed on branch
Rebasing (3/3)
9ae6f48a90e9eaec8d1bcc25b249c4809e2e3f4e
Fast-forward to 9ae6f48a90e9eaec8d1bcc25b249c4809e2e3f4e
HEAD is now at 9ae6f48 merged branch into master
Successfully rebased and updated refs/heads/master.
```

Result with verbose on, `git commit --no-edit --amend --all` included:
```
$ git rebase --continue
Rebasing (2/3)
6c0b5a30e15ce0e463a22414de42a3258688e64d
Fast-forward to 6c0b5a30e15ce0e463a22414de42a3258688e64d
HEAD is now at 6c0b5a3 changed on branch
Rebasing (3/3)
6ab88e7136306794c96ea1ee4abf7fb451c86a83
fatal: Unknown option for merge-recursive: -X'theirs'
Error redoing merge 6ab88e7136306794c96ea1ee4abf7fb451c86a83
```

The above outputs where generated with `git version 2.7.4 (Apple Git-66)`

I would prefer to change the tree instead of just the commit, such that the merge needs to be redone even if git rebase is smart enough to reuse merged tree when parent trees are the same.

> @sonntam Could you prepare a pull request, please?

I think this fix needs to go upstream, so I will need to send a patch to the mailing list.

**Comment by dscho** [comment]
> I think this fix needs to go upstream, so I will need to send a patch to the mailing list.

I concur that this should go upstream, and I thank you so much for preparing the patch submission.

**Comment by dscho** [comment]
Y'all.

This was already almost done, and it is a bit unfair to the people who do all the work to expect them to do *all* the work.

As a consequence, you did not get what you want.

Let this be a strong hint that y'all can do a lot better. Rather than sitting by idly, engage. Then you get what you want.

Y'all are lucky that I just stumbled over a similar bug report on the Git mailing list, and remembered this here ticket. So the patch [got submitted](https://public-inbox.org/git/8cafe900ce97d885c58f4a7f02e748d2ae3b02a2.1515101305.git.johannes.schindelin@gmx.de).

But again, you do not have to wait for others to do all the work. You can be active. You can be the masters of your fate.

**Comment by kgybels** [comment]
@dscho Thank you for picking up the slack. I just confirmed that the bug is fixed in v2.16.0-rc2.windows.1. We can close this issue, or will it be automatically closed when the commit reaches a certain branch?

**Comment by dscho** [comment]
I think we can close it ;-)
